docs/keps/20200504-cleanup-fed-resources.md

```diff
@@ -0,0 +1,113 @@
+* Define how users can indicate that a federated resource should be removed when a cluster leaves federation
+* Define removal strategies
+
+## Proposals
+
+The APIResource struct should be updated to contain a `RemoveStrategy` member. `RemoveStrategy` will initially support two values, `BestEffort` and `Required`.
+
+  * The `BestEffort` strategy will **not** prevent a cluster from unjoining when there are errors removing the resource.
+  * The `Required` strategy will halt the unjoin process and set the KubefedCluster resource in a failed state when encountering errors. Unjoining is blocked until all resources with the `Required` `RemoveStrategy` have been removed successfully.
+
+By default, `RemoveStrategy` will be nil on all federated types. A user must explicitly set a removal strategy on the resource they are creating. Types without a removal strategy will not be removed.
+
+
+### Implementation details
+
+* Update APIResource to contain a new member:
+  ```
+    // RemoveStrategy indicates what strategy should be employed when removing
+    // federated resources during nnjoin
+    RemoveStrategy *RemoveStrategy `json:"removeStrategy,omitempty"`
+  ```
+
+  ```
+    type RemoveStrategy string
+
+    const (
+      RemoveStrategyBestEffort RemoveStrategy = "BestEffort"
+      RemoveStrategyRequired   RemoveStrategy = "Required"
+    )
+  ```
+* When labels are applied to federated resources (ApplyOverrides), add the following label if RemoveStrategy is !nil:
+  * `kubefed.io/remove: best-effort || required`
+* During unjoin, query for managed (`kubefed.io/managed: true`) resources with the `kubefed.io/remove` label: `[{ key: "kubefed.io/remove", operator: "Exists" }]`
+
+#### Best Effort Strategy
+
+Errors or timeouts that occur while removing resources with the best effort removal strategy should be logged but should not halt the unjoining process.
+
+#### Required Strategy
+
+Resources with the "required" removal strategy must be removed successfully on unjoin. After removing the resource, the client must verify that the resource is no longer present. If a resource cannot be removed:
+
+* UnjoinCluster should return a formatted error containing a list of resources which could not be removed
+* A ClusterCondition should be added to the KubefedCluster status with a message indicating which resources could not be removed
+* (optional) Set the ConditionType to `Failed`
+
+
+### Additional ClusterConditionTypes (optional)
+
+Presently, there are only two ClusterConditionTypes, `Ready` and `Offline`. Given the `required` remove strategy, a `KubefedCluster` may end up in a state where the cluster is `Ready` but is in the process of "Unjoining". Adding additional conditions will make it easier to communicate federation life cycle status to the cluster operator.
+
+#### Unjoining ClusterConditionType
+
+In continuation of the existing nomenclature, the `Unjoining` condition should be set on a cluster when the operated initiates `UnjoinCluster`.
```

Comment:
Please, explain how the process would work. It would then exist a condition `UnjoinFailed`.

Reply:
I agree. It would probably best to list the possible scenarios:

Ready => Unjoining => Offline
Ready => Unjoining => UnjoinFailed

In the latter case we should also specify that an appropriate reason/message is set on the condition.